src/Uno.UI/UI/Xaml/VisualStateGroup.cs

```diff
@@ -262,78 +239,136 @@ internal void GoToState(IFrameworkElement element, VisualState state, VisualStat
+			}
+
+			// Finally effectively apply the target state!
+			if (useTransitions && target.transition?.Storyboard is { } transitionAnimation)
 			{
-				ResourceResolver.PushNewScope(_xamlScope);
+				// Note: As of 2021-08-16 win 19043, if the transitionAnimation is Repeat=Forever, we actually never apply the state!
 
-				this.CurrentState = state;
-				if (this.CurrentState != null && element != null)
+				transitionAnimation.Completed += OnTransitionCompleted;
+				transitionAnimation.Begin();
+
+				void OnTransitionCompleted(object s, object a)
 				{
-					foreach (var setter in this.CurrentState.Setters.OfType<Setter>())
+					transitionAnimation.Completed -= OnTransitionCompleted;
+
+					if (target.state?.Storyboard is { } stateAnimation)
 					{
-						setter.ApplyValue(DependencyPropertyValuePrecedences.Animations, element);
+						transitionAnimation.TurnOverAnimationsTo(stateAnimation);
 					}
+
+					ApplyTargetState();
 				}
+			}
+			else
+			{
+				ApplyTargetState();
+			}
 
-				if (transition?.Storyboard == null || !useTransitions)
+			void ApplyTargetState()
+			{
+				// Apply target state setters (the right time to do it!) 
+				if (!FeatureConfiguration.VisualState.ApplySettersBeforeTransition)
 				{
-					onTransitionComplete(this, null);
+					ApplyTargetStateSetters();
+				}
+
+				// Starts target state animation
+				if (target.state?.Storyboard is { } stateAnimation)
+				{
+					stateAnimation.Completed += OnStateStoryboardCompleted;
+					stateAnimation.Begin();
+
+					void OnStateStoryboardCompleted(object s, object a)
+					{
+						state.Storyboard.Completed -= OnStateStoryboardCompleted;
+						onStateChanged();
+					}
 				}
 				else
 				{
-					transition.Storyboard.Completed += onTransitionComplete;
-					transition.Storyboard.Begin();
+					onStateChanged();
 				}
 			}
+
+			void ApplyTargetStateSetters()
+			{
+				if (target.state is null || element is null)
+				{
+					return;
+				}
+
 #if !HAS_EXPENSIVE_TRYFINALLY
-			finally
+				try
 #endif
-			{
-				ResourceResolver.PopScope();
+				{
+					ResourceResolver.PushNewScope(_xamlScope);
+
+					foreach (var setter in target.state.Setters.OfType<Setter>())
+					{
+						setter.ApplyValue(DependencyPropertyValuePrecedences.Animations, element);
+					}
+				}
+#if !HAS_EXPENSIVE_TRYFINALLY
+				finally
+#endif
+				{
+					ResourceResolver.PopScope();
+				}
 			}
 		}
 
 		private VisualTransition FindTransition(string oldStateName, string newStateName)
 		{
-			if (oldStateName.IsNullOrEmpty() || newStateName.IsNullOrEmpty())
-			{
-				return null;
-			}
-
-			var perfectMatch = Transitions.FirstOrDefault(vt =>
-				string.Equals(vt.From, oldStateName) &&
-				string.Equals(vt.To, newStateName));
+			var hasOld = oldStateName.HasValue();
+			var hasNew = newStateName.HasValue();
 
-			if (perfectMatch != null)
+			if (hasOld && hasNew && Transitions.FirstOrDefault(Match(oldStateName, newStateName)) is { } perfectMatch)
```

Comment:
a comment would be nice here to explain the complexity of the logic (which is unnecessary at first glance)